apps/sales/views.py

```diff
@@ -1457,6 +1457,428 @@ def put(self, request, pk, *args, **kwargs):
+        return self.update(request, *args, **kwargs)
+    
+    @transaction.atomic
+    def update(self, request, pk, *args, **kwargs):
+
+        # ----------------------------------- D A T A  V A L I D A T I O N -----------------------------#
+        """
+        All the data in request will be validated here. it will handle the following errors:
+        - Invalid data types
+        - Invalid foreign keys
+        - nulls in required fields
+        """
+        # Get the given data from request
+        given_data = request.data
+        errors = {}
+
+        # Validate SaleReceipt Data
+        sale_receipt_data = given_data.pop('sale_receipt', None)  # parent_data
+        if sale_receipt_data:
+            sale_receipt_data['sale_receipt_id'] = pk
+            sale_receipt_error = validate_multiple_data(self, [sale_receipt_data], SaleReceiptSerializer, [])
+            if sale_receipt_error:
+                errors["sale_receipt"] = sale_receipt_error
+
+        if errors:
+            return build_response(0, "ValidationError :", errors, status.HTTP_400_BAD_REQUEST)
+
+        # ------------------------------ D A T A   U P D A T I O N -----------------------------------------#
+        custom_data = {
+            'sale_receipt':{}
+            }
+        
+        '''
+        This "save_assignments_history_with_end_date" function saves the instance in 'LeadAssignmentHistory' model
+        by updating the end date.
+
+        def save_assignments_history_with_end_date(self):
+            now = datetime.now()
+            end_date = now.strftime('%Y-%m-%d %H:%M:%S')
+            history_data_set = LeadAssignmentHistory.objects.filter(sale_receipt_id=pk).order_by('created_at').last()
+            history_data_set.end_date = end_date # update with end date
+            history_data_set.save() # save the record
+            logger.info(f'last history_id : {history_data_set}')
+
+
+        This "create_new_history" function creates the new instance in 'LeadAssignmentHistory' model
+        with required fields to be updated.
+
+        def create_new_history(self,update_fields):
+            assignment_history = {}
+            assignments_history = generic_data_creation(self, [assignment_history], LeadAssignmentHistorySerializer, update_fields)
+            assignments_history = assignments_history if assignments_history else []
+            logger.info('LeadAssignmentHistory - created*') '''
+
+        if sale_receipt_data:
+            '''
+            This block performs updation of end date in case of 'assignee_id' and 'sale_receipt_id' is changed.
+            (detects changes in 'SaleReceipt' payload)
+            Adds new record in 'LeadAssignmentHistory' if 'assignee_id' in 'SaleReceipt' is changed
+            by leaving end date as null.
+            Updates old record in 'LeadAssignmentHistory' if 'sale_receipt_id' in 'SaleReceipt' is changed
+            by adding end date
+            '''
```

Comment:
Correct above comments, what is LeadAssignmentHistory here?

Reply:
Removed Thulasi